### PR TITLE
MAT-293 – support over-refunds

### PR DESCRIPTION
**Pull request by NoelLH**
And tidy up our #stripe Slack notification config a bit

### CODE REVIEW

File: tests/Application/Actions/Hooks/StripeChargeUpdateTest.php [hunk 1]
```diff
@@ -367,6 +372,64 @@ public function testSuccessfulFullRefund(): void
...
+
+        $request = $this->createRequest('POST', '/hooks/stripe', $body)
+            ->withHeader('Stripe-Signature', $this->generateSignature($time, $body, $webhookSecret));
+
+        $response = $app->handle($request);
+
+        $this->assertEquals('ch_externalId_123', $donation->getChargeId());
+        $this->assertEquals(DonationStatus::Refunded, $donation->getDonationStatus());
+        $this->assertEquals('1.00', $donation->getTipAmount());
+        $this->assertEquals(200, $response->getStatusCode());
```
Comment (bdsl) [inline tests/Application/Actions/Hooks/StripeChargeUpdateTest.php:408]: this test has now got quite long at imho at 42 lines, its hard to see exactly what it's testing - especially when the name is just `testSuccessfulFullRefund`. May be worth splitting it up into multiple tests, extracting functions, or even just breaking it up into with given/when/then or arrange/act/assert comments?

Comment (bdsl) [reply tests/Application/Actions/Hooks/StripeChargeUpdateTest.php:408]: I could comment on details above but probably not worth adding more before I get your thoughts on the first comment @NoelLH . Happy to pair on splitting it / making it clearer.

File: tests/Application/Actions/Hooks/StripeChargeUpdateTest.php [hunk 2]
```diff
@@ -367,6 +372,64 @@ public function testSuccessfulFullRefund(): void
...
+    }
+
+    public function testSuccessfulOverRefund(): void
```
Comment (bdsl) [inline tests/Application/Actions/Hooks/StripeChargeUpdateTest.php:411]: `Successful` feels like a bit of a meaninguless word to me. Would it be possible to make the test name a full sentence that describes the main thing we're asserting? Looking at the test I think it might be `testOverRefundSendsSlackNoticeAndUpdatesRecord`. But the fact that there's an `and` in there makes me think maybe it should be two tests.

File: tests/Application/Actions/Hooks/StripeChargeUpdateTest.php [hunk 3]
```diff
@@ -452,7 +516,15 @@ public function testUnsupportedRefundAmount(): void
 
         $body = $this->getStripeHookMock('ch_unsupported_partial_refund');
         $donation = $this->getTestDonation();
-        $webhookSecret = $container->get('settings')['stripe']['accountWebhookSecret'];
+        /**
+         * @var array{
+         *    stripe: array{
+         *      accountWebhookSecret: string
```
Comment (bdsl) [inline tests/Application/Actions/Hooks/StripeChargeUpdateTest.php:522]: The need to repeat this and similar makes me think it might be worth changing the value of 'settings' in the container from an array to an object with an actual class. Could be read-only with all public properties. Not a blocker here, but might be worth doing before repeating this sort of thing too many times. Using `@var` is error-prone, there's no check that the var actually hold what we say it holds.

File: app/dependencies.php [hunk 4]
```diff
@@ -91,9 +92,26 @@ function (ContainerInterface $c): Auth\DonationPublicAuthMiddleware {
...
+        StripeChatterInterface::class => static function (ContainerInterface $c): ChatterInterface {
+            /**
+             * @var array{
+             *    notifier: array{
+             *      slack: array{
+             *        stripe_channel: string
+             *      }
+             *    }
+             *  } $settings
+             */
```
Comment (bdsl) [inline app/dependencies.php:104]: See https://github.com/thebiggive/matchbot/pull/537/files#r1192272402

File: app/dependencies.php [hunk 5]
```diff
@@ -91,9 +92,26 @@ function (ContainerInterface $c): Auth\DonationPublicAuthMiddleware {
...
+            $settings = $c->get('settings');
+            $stripeChannel = $settings['notifier']['slack']['stripe_channel'];
+            /** @var SlackChannelChatterFactory $chatterFactory */
+            $chatterFactory = $c->get(SlackChannelChatterFactory::class);
+            return $chatterFactory->makeChatter($stripeChannel);
+        },
+
         SlackChannelChatterFactory::class => static function (ContainerInterface $c): SlackChannelChatterFactory {
+            /** @var array $settings */
             $settings = $c->get('settings');
```
Comment (bdsl) [inline app/dependencies.php:114]: Why the change from `assert` to `@var`? Seem to be going in the wrong direction, away from type safety.

File: src/Application/Actions/Hooks/StripeChargeUpdate.php [hunk 6]
```diff
@@ -182,6 +189,18 @@ private function handleChargeDisputeClosed(Event $event, Response $response): Re
             return $this->respond($response, new ActionPayload(204));
         }
 
+        if ($donation->getAmountFractionalIncTip() < $dispute->amount) {
```
Comment (bdsl) [inline src/Application/Actions/Hooks/StripeChargeUpdate.php:192]: minor point but since we have more than in the comments imho it's easier a bit easier to follow if we use the more than symbol in the code too?
```suggestion
        if ($dispute->amount > $donation->getAmountFractionalIncTip()) {
```

File: src/Application/Actions/Hooks/StripeChargeUpdate.php [hunk 7]
```diff
@@ -214,6 +234,14 @@ private function handleChargeRefunded(Event $event, Response $response): Respons
 
         $isTipRefund = $donation->getTipAmountFractional() === $amountRefunded;
         $isFullRefund = $donation->getAmountFractionalIncTip() === $amountRefunded;
+        // If things disagree about the original captured amount, this is still uncharted
+        // territory and we should refrain from processing. If that matches but there is an
+        // over-refund, we'll want to notify the team but we know we are best off handling it
+        // like a simple refund from a donation status perspective.
+        $isOverRefund = (
+            $donation->getAmountFractionalIncTip() === $charge->amount_captured &&
+            $donation->getAmountFractionalIncTip() < $amountRefunded
```
Comment (bdsl) [inline src/Application/Actions/Hooks/StripeChargeUpdate.php:243]: IMHO easier to read this way - we think of it as an over-refund, not an under-donation:
```suggestion
             $amountRefunded > $donation->getAmountFractionalIncTip()
```

Comment (bdsl) [inline src/Application/Actions/Hooks/StripeChargeUpdate.php:243]: IMHO easier to read this way - we think of it as an over-refund, not an under-donation:
```suggestion
             $amountRefunded > $donation->getAmountFractionalIncTip()
```

File: tests/Application/Actions/Hooks/StripeChargeUpdateTest.php (hunk 3 above)
Comment (NoelLH) [reply tests/Application/Actions/Hooks/StripeChargeUpdateTest.php:522]: @bdsl I think we talked about this a couple of weeks ago. The latest Slim Skeleton and newer apps do this so we have a model to follow, but it's not a super quick change.

File: app/dependencies.php (hunk 5 above)
Comment (NoelLH) [reply app/dependencies.php:114]: I guess no great reason except that it was closer to the other psalm-placating changes I was thinking about, and as discussed `$settings` being a generic array will likely be temporary anyway.

File: tests/Application/Actions/Hooks/StripeChargeUpdateTest.php (hunk 2 above)
Comment (NoelLH) [reply tests/Application/Actions/Hooks/StripeChargeUpdateTest.php:411]: For now I've given it this name and added a comment. I think that releasing match funds is important too – adding this test helped me spot that the logic for that needed to change.

File: src/Application/Actions/Hooks/StripeChargeUpdate.php [hunk 8]
```diff
@@ -182,6 +189,18 @@ private function handleChargeDisputeClosed(Event $event, Response $response): Re
...
+        if ($donation->getAmountFractionalIncTip() < $dispute->amount) {
+            // More than the original amount was returned/refunded, which we don't
+            // *really* expect but are best to continue processing as a full refund
+            // with a warning.
+            $this->warnAboutOverRefund(
+                'charge.dispute.closed (lost)',
+                $donation,
+                $dispute->amount,
+                $dispute->currency,
+            );
```
[the remark this answers is not in the document]
Comment (NoelLH) [reply src/Application/Actions/Hooks/StripeChargeUpdate.php:201]: Yep true – I'm working on this

File: src/Application/Actions/Hooks/StripeChargeUpdate.php [hunk 9]
```diff
@@ -249,11 +291,48 @@ private function handleChargeRefunded(Event $event, Response $response): Respons
...
 
         return $this->respondWithData($response, $charge);
     }
 
+    /**
+     * @param string $eventType
+     * @param Donation $donation
+     * @param int $refundedOrDisputedAmount In small currency unit, e.g. pence or cents.
+     * @param string $refundedOrDisputedCurrencyCode
+     */
```
Comment (bdsl) [inline src/Application/Actions/Hooks/StripeChargeUpdate.php:304]: IMHO easier to see the important parts if we leave out the docblock part that jsut repeats what's in the code:
```suggestion
     * @param int $refundedOrDisputedAmount In small currency unit, e.g. pence or cents.
     */
```

I appreciate the specification that $refundedOrDisputedAmount is in a small currency unit - but also the fact that we need to specify it makes me think it'd would be good to introduce a value object type and use that instead of `int`. I'd probably combine the amount with the currency So the calling code could look something like:

```php
            $this->warnAboutOverRefund(
                'charge.dispute.closed (lost)',
                $donation,
                Money::of(currencySymbol: $dispute->currency, minorUnits: $dispute->amount)
            );
```

File: src/Application/Actions/Hooks/StripeChargeUpdate.php [hunk 10]
```diff
@@ -214,6 +234,14 @@ private function handleChargeRefunded(Event $event, Response $response): Respons
...
+        // territory and we should refrain from processing. If that matches but there is an
+        // over-refund, we'll want to notify the team but we know we are best off handling it
+        // like a simple refund from a donation status perspective.
+        $isOverRefund = (
+            $donation->getAmountFractionalIncTip() === $charge->amount_captured &&
+            $donation->getAmountFractionalIncTip() < $amountRefunded
+        );
 
         // Available status' (pending, succeeded, failed, canceled),
         // see: https://stripe.com/docs/api/refunds/object.
```
[the remark this answers is not in the document]
Comment (NoelLH) [reply src/Application/Actions/Hooks/StripeChargeUpdate.php:247]: I guess we don't – will update

File: src/Application/Notifier/StripeChatterInterface.php [hunk 11]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+namespace MatchBot\Application\Notifier;
+
+use Symfony\Component\Notifier\ChatterInterface;
+
+interface StripeChatterInterface extends ChatterInterface
+{}
```
Comment (bdsl) [inline src/Application/Notifier/StripeChatterInterface.php:8]: Feels strange to extend the interface without adding anything. This is just so that the container can distinguish the two instances and know where to inject each one right? Not a blocker but I feel like there should be a nicer way to do that.

And actually if we are gong the way of extending it don't we really need to give it two children - one `StripeChatterInterface` and one `MatchbotChatterInterface`? Otherwise it looks like anything that declares a dependency on the parent interface doesn't care which channel its messages go to.

File: src/Application/Actions/Hooks/StripeChargeUpdate.php (hunk 9 above)
Comment (bdsl) [reply src/Application/Actions/Hooks/StripeChargeUpdate.php:304]: on the other hand as this is a private function it's not a big issue to manually check that the usage semantics are correct. Just a more general point.

File: src/Application/Notifier/StripeChatterInterface.php (hunk 11 above)
Comment (NoelLH) [reply src/Application/Notifier/StripeChatterInterface.php:8]: Yes, that's why. I think it's OK for consumers to essentially "not care" for now, it's not the end of the world if we have a default channel that doesn't require much thought. I'd be interested in other options than extending the interface but couldn't immediately think of one that would work well with Slim DI. This felt like an improvement on essentially doing DI stuff inside the payout handler as we were before.

File: tests/Application/Actions/Hooks/StripeChargeUpdateTest.php (hunk 1 above)
Comment (NoelLH) [reply tests/Application/Actions/Hooks/StripeChargeUpdateTest.php:468]: I've added grouping comments for now.

File: tests/Application/Actions/Hooks/StripeChargeUpdateTest.php [hunk 12]
```diff
@@ -367,6 +372,70 @@ public function testSuccessfulFullRefund(): void
         $webhookSecret = $container->get('settings')['stripe']['accountWebhookSecret'];
         $time = (string) time();
 
+        $chatterProphecy = $this->prophesize(StripeChatterInterface::class);
+        // Double-check that the normal success case isn't messaging Slack.
+        $chatterProphecy->send(Argument::type(ChatMessage::class))->shouldNotBeCalled();
```
Comment (bdsl) [inline tests/Application/Actions/Hooks/StripeChargeUpdateTest.php:434]: Can make a slightly stronger / more general assertion.
```suggestion
        $chatterProphecy->send(Argument::cetera())->shouldNotBeCalled();
```
I.e. we don't care what the arguments are or aren't, this method should just not be called at all.

File: src/Application/Notifier/StripeChatterInterface.php (hunk 11 above)
Comment (bdsl) [reply src/Application/Notifier/StripeChatterInterface.php:8]: It's not that we shouldn't have a default channel, it's that the the consumers don't specify that they want the default channel. If the parent interface promises to give them a default channel then to be a Liskov-subtype the child interface should give them the default channel too. Can leave it for now, although I think making the third interface would make it easier to understand.

File: tests/Application/Actions/Hooks/StripeChargeUpdateTest.php [hunk 13]
```diff
@@ -367,6 +428,73 @@ public function testSuccessfulFullRefund(): void
...
+        $this->assertEquals(DonationStatus::Refunded, $donation->getDonationStatus());
+        $this->assertEquals('1.00', $donation->getTipAmount());
+        $this->assertEquals(200, $response->getStatusCode());
+    }
+
+    /**
+     * Verifies that:
+     * * Slack gets a warning with the correct details
+     * * The donation record is updated with the new status, and
+     * * Match funds are released.
```
Comment (bdsl) [inline tests/Application/Actions/Hooks/StripeChargeUpdateTest.php:475]: :+1:

File: tests/Application/Actions/Hooks/StripeChargeUpdateTest.php [hunk 14]
```diff
@@ -367,6 +428,73 @@ public function testSuccessfulFullRefund(): void
...
+     */
+    public function testOverRefundSendsSlackNoticeAndUpdatesRecord(): void
+    {
+        // arrange
+        $app = $this->getAppInstance();
+        /** @var Container $container */
+        $container = $app->getContainer();
```
Comment (bdsl) [inline tests/Application/Actions/Hooks/StripeChargeUpdateTest.php:482]: Probably not worth changing now, but I wonder if using the container and the app instance actually gives us much in a unit test? I feel like its sometimes easier to see exactly what the code being tested is if the test directly instantiates e.g. the `StripeChargeUpdate` class and passes in the dependencies.  Or if we want to test the app as a whole maybe go the other way and do an integration test with the real DB then we test the repository code too.

File: tests/Application/Actions/Hooks/StripeChargeUpdateTest.php [hunk 15]
```diff
@@ -367,6 +428,73 @@ public function testSuccessfulFullRefund(): void
...
+        $time = (string) time();
+
+        $chatterProphecy = $this->prophesize(StripeChatterInterface::class);
+
+        $options = (new SlackOptions())
+            ->block((new SlackHeaderBlock('[test] Over-refund detected')))
+            ->block((new SlackSectionBlock())->text('Over-refund detected for donation 12345678-1234-1234-1234-1234567890ab based on charge.refunded hook. Donation inc. tip was 124.45 GBP and refund or dispute was 134.45 GBP'))
+            ->iconEmoji(':o');
+        $message = (new ChatMessage('Over-refund detected'))
+            ->options($options);
```
Comment (bdsl) [inline tests/Application/Actions/Hooks/StripeChargeUpdateTest.php:496]: Could rename `$message` to `$expectedMessage` to make it clearer - but I think even better is just to inline all this stuff so it appears only in the assert section, e.g.:

```php
        $chatterProphecy->send((new ChatMessage('Over-refund detected'))
            ->options((new SlackOptions())
                ->block((new SlackHeaderBlock('[test] Over-refund detected')))
                ->block((new SlackSectionBlock())->text('Over-refund detected for donation 12345678-1234-1234-1234-1234567890ab based on charge.refunded hook. Donation inc. tip was 124.45 GBP and refund or dispute was 134.45 GBP'))
                ->iconEmoji(':o')))->shouldBeCalledOnce();
```

Also TIL that prophecy will let you do `shouldBeCalled` after the the action, using the revelation as a spy instead of a mock. Nice!

Comment (NoelLH) [reply tests/Application/Actions/Hooks/StripeChargeUpdateTest.php:496]: Yeah, in theory that would be nicer. I tried moving some of the calls to `// assert` but of course if they need to define things about the prophesised method's behaviour in the same chained call then to do it with the fluent interface you need both things before you `// act`

**Comment by bdsl** [comment]
Initial review done, all LGTM other than where I've commented. Just going through updates now.

**Review by bdsl — APPROVED** [review]
LGTM - thanks for the adjustments.

File: tests/Application/Actions/Hooks/StripeChargeUpdateTest.php (hunk 15 above)
Comment (NoelLH) [reply tests/Application/Actions/Hooks/StripeChargeUpdateTest.php]: I think it was moving donation repo calls at the same time which broke things, but I didn't dig into it further
